tensorboard/components/tf_ng_tensorboard/core/core.reducers.ts

```diff
@@ -29,30 +33,85 @@ import * as actions from './core.actions';
 
 export const CORE_FEATURE_KEY = 'core';
 
+export interface LoadState {
+  state: DataLoadState;
+  // Time since epoch.
+  lastLoadedTimeInMs: number | null;
+}
+
 export interface CoreState {
   activePlugin: PluginId | null;
   plugins: PluginsListing;
+  pluginsListLoaded: LoadState;
+  reloadPeriodInMs: number;
+  reloadEnabled: boolean;
 }
 
 export interface State {
-  [CORE_FEATURE_KEY]: CoreState;
+  [CORE_FEATURE_KEY]?: CoreState;
 }
 
 const initialState: CoreState = {
   activePlugin: null,
   plugins: {},
+  pluginsListLoaded: {
+    state: DataLoadState.NOT_LOADED,
+    lastLoadedTimeInMs: null,
+  },
+  reloadPeriodInMs: 30000,
+  reloadEnabled: true,
 };
 
 const reducer = createReducer(
   initialState,
   on(actions.changePlugin, (state: CoreState, {plugin}) => {
     return {...state, activePlugin: plugin};
   }),
+  on(actions.pluginsListingRequested, (state: CoreState) => {
+    return {
+      ...state,
+      pluginsListLoaded: {
+        ...state.pluginsListLoaded,
+        state: DataLoadState.LOADING,
+      },
+    };
+  }),
+  on(actions.pluginsListingFailed, (state: CoreState) => {
+    return {
+      ...state,
+      pluginsListLoaded: {
+        ...state.pluginsListLoaded,
+        state: DataLoadState.FAILED,
+      },
+    };
+  }),
   on(actions.pluginsListingLoaded, (state: CoreState, {plugins}) => {
     const [firstPlugin] = Object.keys(plugins);
     let activePlugin =
       state.activePlugin !== null ? state.activePlugin : firstPlugin;
-    return {activePlugin, plugins};
+    return {
+      ...state,
+      activePlugin,
+      plugins,
+      pluginsListLoaded: {
+        state: DataLoadState.LOADED,
+        lastLoadedTimeInMs: Date.now(),
+      },
+    };
+  }),
+  on(actions.toggleReload, (state: CoreState) => {
+    return {
+      ...state,
+      reloadEnabled: !state.reloadEnabled,
+    };
+  }),
+  on(actions.changeReloadPeriod, (state: CoreState, {periodInMs}) => {
+    const nextReloadPeriod =
+      periodInMs > 0 ? periodInMs : state.reloadPeriodInMs;
```

Comment:
Can we treat period = 0 as "never reload"?  (if it's not controversial)

Reply:
When user toggles enable/disable reload, I hope it to remember this value, so no.

Reply:
That makes more sense